### DISCUSSION TITLE
Introduce "task dependency" into pnpm's workspace runner

### DISCUSSION THREAD

**Discussion by kenotron** [pnpm/pnpm#3982] [Ideas]
[`lage`](https://microsoft.github.io/lage/) and [`turbo`](https://turborepo.com) both support pipelining via a separate "task dependency" definition. I see that we're beginning to build workspace-wide task running into package (workspace) managers like `pnpm`. Having this information allows us all to achieve really amazing speeds in build/test/bundle, etc.

See this page for some of the capabilities that one can have: 
https://microsoft.github.io/lage/guide/levels.html

Now, I wouldn't expect `pnpm` to implement all this, but I do wonder if some of the logic or even the configuration can be standardized. Say: how do we do incremental build caches (which can actually replace TS proj references), scoping builds, and pipelining.

**Comment by zkochan** [comment]
pnpm might be able to more effectively download the cache

**Comment by zkochan** [comment]
I have tried to setup lage in the pnpm monorepo and there is an issue right from the start. It cannot work with a cycle in the dependencies (even if the cycle is through devDependencies).

**Comment by gluxon** [comment]
`lage` and `turbo` are both really interesting. Thanks for linking to them.

Just to add another data point, my org's team chose [Gradle](https://gradle.org/) to achieve package-parallel TypeScript, babel, webpack builds. I think the killer feature there is Gradle's [build scan](https://blog.gradle.org/introducing-build-scans) feature, which allows you to dive into unexpected cache misses/hits with a web GUI. It's really nice to get a web UI telling you *"Hey, we had to rebuild this package because this dependency in `node_modules` was missing (or a different version)."*

I think the "build scan" feature of Gradle is what `lage`, `turbo`, and [`rush`](https://rushjs.io/) appear to be behind on. I suspect `turbo` or `rush` will implement that sooner or later, which makes me think this sort of functionality might be a feature creep for pnpm? I worry about wasted effort in pnpm to catch up to tools that are getting more dedicated development in this area.
